Content.Server/Chat/V2/Repository/ChatRepository.cs

```diff
@@ -0,0 +1,202 @@
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Content.Shared.Chat.V2;
+using Content.Shared.Chat.V2.Repository;
+using Robust.Server.Player;
+using Robust.Shared.Network;
+using Robust.Shared.Replays;
+
+namespace Content.Server.Chat.V2.Repository;
+
+/// <summary>
+/// Stores <see cref="IChatEvent"/>, gives them UIDs, and issues <see cref="MessageCreatedEvent"/>.
+/// Allows for deletion of messages.
+/// </summary>
+public sealed class ChatRepositorySystem : EntitySystem
+{
+    [Dependency] private readonly IReplayRecordingManager _replay = default!;
+    [Dependency] private readonly IPlayerManager _player = default!;
+
+    // Clocks should start at 1, as 0 indicates "clock not set" or "clock forgotten to be set by bad programmer".
+    private uint _nextMessageId = 1;
+    private Dictionary<uint, ChatRecord> _messages = new();
+    private Dictionary<NetUserId, List<uint>> _playerMessages = new();
+
+    public override void Initialize()
+    {
+        Refresh();
+
+        _replay.RecordingFinished += _ =>
+        {
+            // TODO: resolve https://github.com/space-wizards/space-station-14/issues/25485 so we can dump the chat to disc.
+            Refresh();
+        };
+    }
+
+    /// <summary>
+    /// Adds an <see cref="IChatEvent"/> to the repo and raises it with a UID for consumption elsewhere.
+    /// </summary>
+    /// <param name="ev">The event to store and raise</param>
+    /// <returns>If storing and raising succeeded.</returns>
+    public bool Add(IChatEvent ev)
+    {
+        if (!_player.TryGetSessionByEntity(ev.Sender, out var session))
+        {
+            return false;
+        }
+
+        var messageId = _nextMessageId;
+
+        _nextMessageId++;
+
+        ev.Id = messageId;
+
+        var storedEv = new ChatRecord
+        {
+            UserName = session.Name,
+            UserId = session.UserId,
+            EntityName = Name(ev.Sender),
+            StoredEvent = ev
+        };
+
+        _messages[messageId] = storedEv;
+
+        if (!_playerMessages.TryGetValue(storedEv.UserId, out var set))
+        {
+            set = new List<uint>();
+            _playerMessages[storedEv.UserId] = set;
+        }
+
```

Comment:
CollectionsMarshal.GetValueRefOrAddDefault() is a faster version of this. It's not sandboxed but sandbox doesn't run on server anyway (and likely won't ever for ss14).